<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<meta name="apple-mobile-web-app-capable" content="yes">

<style>
@import url(CSS/stack.css);
@import url(CSS/d3-tip.css);
@import url(CSS/figures.css);
@import url(CSS/kbroman_talk.css);
@import url(CSS/corr_w_scatter.css);
/* comment following line to get rid of notes; uncommented to show notes */
/* @import url(CSS/kbroman_presentation.css); */
</style>

<script charset="utf-8" type="text/javascript" src="JS/d3.min.js"></script>
<script type="text/javascript" src="JS/d3.tip.js"></script>
<script type="text/javascript" src="JS/slide_transition.js"></script>


<title>Interactive graphics for high-dimensional genetic data</title>
</head>
<body class="overview">



<section class="titlepage">
<h3 style="margin-top:-10px;" class="highlight">Interactive graphics
for high-dimensional genetic data</h2>

<h4><a href="http://www.biostat.wisc.edu/~kbroman">Karl Broman</a></h4>
<p class="nopadding">Biostatistics &amp; Medical Informatics, University of Wisconsin
&ndash; Madison</p>

<p class="nopadding"><code>
<a href="http://www.biostat.wisc.edu/~kbroman" class="gray">biostat.wisc.edu/~kbroman</a><br>
<a href="https://github.com/kbroman" class="gray">github.com/kbroman</a><br>
<a href="https://twitter.com/kwbroman" class="gray">@kwbroman</a>
</code></p>

<p class="codelink"><a href="presentation.html" class="gray">Remove stickies</a></p>

<aside>
These are slides for a talk for the 4th Symposium on Biological Data
Visualization (BioVis), in Boston, Massachusetts, 12 July 2014.<br><br>

Click on the link at the bottom-left to remove these
notes.<br><br>

These slides were tested for Chrome on a Mac and might look terrible
in some browsers.
</aside>

</section>



<section>
<h3 class="abscenter">applied statistician</h3>
<aside>
I&rsquo;m an applied statistician, though I&rsquo;d prefer to drop the
&ldquo;applied&rdquo; as if it&rsquo;s not applied, it&rsquo;s not
statistics.<br><br>

My job is to help scientists make sense of data. Data visualization
is always a big part of that.<br><br>

(The next slide is a gratuitous animation. Teach a person a bit of
JavaScript, and you&rsquo;ll get gratuitous animations as a result.)
</aside> <p class="pg">2</p></section>

<section>
<h3 class="abscenter"><font id="fadeout_applied">applied</font> statistician</h3>
<p class="pg">2</p></section>






<section>

<h3>Inbred mice</h3>

<img style="position:fixed;bottom: 0px; left:0px; z-index: -1;" width="1280px" src="Figs/inbredmice.jpg" />

<aside>
I focus on genetics problems, and particularly on mouse genetics.<br><br>

I think these are <a
href="http://jaxmice.jax.org/strain/000689.html">SWR mice</a>; the
photo is from <a
href="http://cals.ncsu.edu/genetics/index.php/people/david-threadgill/">David
Threadgill</a>.
</aside>

<p class="pg">3</p></section>






<section>

<h3>Human vs mouse</h3>

<img style="position:fixed;top: 300px; left:120px; z-index: -1;" height="500px" src="Figs/da-vinci-man.jpg" />

<img style="position:fixed;top: 300px; right:120px; z-index: -1;" height="500px" src="Figs/vitruvian_mouse.jpg" />

<aside>
Mice are not humans, but you can learn a great deal about human
biology and disease from mice.<br><br>

The figure on the right is from <a
href="http://www.daviddeen.com">David Deen</a>.
</aside>
<p class="pg">4</p></section>






<section>

<h3>Intercross</h3>

<img src="Figs/intercross.png" />

<aside>
The main problem I tackle is to try to identify genes that underlie
phenotypic variation.<br><br>

The basic experiment I consider is an experimental cross: we start
with two inbred strains that differ in the phenotype and perform a
couple of generations of breeding to mix up the genomes.<br><br>

We gather phenotype data on the progeny, along with genome-wide genotype
data, and we seek to identify genomic regions for which genotype is
associated with phenotype. We call such regions quantitative trait
loci (QTL): genetic loci that influence a quantitative trait.
</aside>
<p class="pg">5</p></section>






<section>

<h3>Genome scan for QTL</h3>

<img src="Figs/lodcurve_insulin.png" />

<aside>
The simplest analysis is to do a statistical test at each locus in the
genome, to assess the association between genotype and
phenotype.<br><br>

We measure the association with a LOD score. LOD scores are log<sub>10</sub>
likelihood ratios &emdash; a transformation of the F statistic from
ANOVA. The details aren&rsquo;t
important. The key is that large LOD scores indicate evidence for an
association.<br><br>

It is important to account for the scan across the genome. The dashed
horizontal line is a threshold for statistical significance. If there
were no association between genotype and phenotype, there&rsquo;d still be a
5% chance of a LOD score that big <font class="highlight">somewhere</font> in the genome.
</aside>
<p class="pg">6</p></section>





<section>

<h3>Genome scan for QTL</h3>

<img src="Figs/lodcurve_insulin_with_effects.png" />

<aside>
Underneath each of these peaks is an association between genotype and
phenotype: If you group the mice by genotype at that location, you&rsquo;ll
see a difference in the average phenotype among the groups.<br><br>

The BB, BR, and RR indicate the three different genotypes, as these
data concern a cross between two strains labeled B and R.
</aside>
<p class="pg">6</p></section>






<section>

<h3>Genome-scale phenotypes</h3>

<img height=700 class="indent" src="Figs/mouse_on_chips.png" />

<aside>
My data visualizations have been pretty pedestrian. But in the past
decade, I&rsquo;ve come to be swamped with data, and I&rsquo;m just beginning to
come to grips with that.<br><br>

This is a picture of a pile of gene expression arrays. More and more,
we&rsquo;re seeing genome-scale phenotype information. For example, in
one of my collaborations, we have data on 500 mice, each with gene
expression microarrays for 6 different tissues.<br><br>

You&rsquo;d think that with such high-dimensional data, we&rsquo;d be
making a lot more graphs. But there&rsquo;s a tendency actually to
make <em>fewer</em> graphs, or even <em>no</em> graphs. It&rsquo;s
hard to look at 500 histograms, so why bother?
</aside>
<p class="pg">7</p></section>





<section>

<h3>Why? &nbsp; <font class="lowlight">summary + details</font></h3>

<div id="manyboxplots">
</div>

<aside>
Visualization of large data sets necessarily involves making a
compressed summary. But with interactive graphs, you can easily
get back to the underlying details.<br><br>

This is like a set of 500 box plots (for the gene expression values
from 500 gene expression microarray), but going farther out into the
tails: the lines connect various quantiles: the 
1, 5, 10, 25, 50, 75, 90, 95, and 99%iles.  <br><br>

Hover over an array above and see the corresponding histogram.  Click
on an array for the histogram to persist; click again to make it go
away.<br><br>

The first 120 or so arrays are badly behaved; they all came from a
common batch that went bad.
</aside>

<p class="codelink">
<a href="Coffee/manyboxplots.coffee" class="gray">code</a> |
<a href="Data/hypo.json" class="gray">data</a></p>

<script type="text/javascript" src="JS/manyboxplots.js"></script>
<p class="pg">8</p></section>






<section>

<h3>Why? &nbsp; <font class="lowlight">summary + details</font></h3>

<div id="corr_w_scatter" style="margin-left:-50px;"></div>

<script type="text/javascript" src="JS/corr_w_scatter.js"></script>

<aside>
Heatmaps of things like correlation matrices can be quite useful, but
they are often hard to read.  It&rsquo;s nice to be able to read off the
values by hovering.<br><br>

Even better: to be able to click on a pixel and view the corresponding
scatterplot.<br><br>

This is a set of genes whose expression values are associated with
genotype at a common locus. In the scatterplot, the points are colored
by genotype at that locus.
</aside>

<p class="codelink">
<a href="Coffee/corr_w_scatter.coffee" class="gray">code</a> |
<a href="Data/corr_w_scatter.json" class="gray">data</a></p>
<p class="pg">9</p></section>



<section nopadding>
<h3><a href="http://d3js.org">D3</a> is awesome!</h3>

<div id="learnd3" class="medtop">
<p>You <font class="darkgray">just</font> need to learn html, css, svg, and javascript.</p>
<p id="dotenter" style="opacity:0;">And don&rsquo;t forget <code>.enter()</code></p>
</div>

<div id="collision" style="position:fixed;top:-10px;left:-50px;width:1280px;height:960px;">
</div>

<div style="position:fixed;bottom:10px;left:10px;">
<code style="font-size:28px;"><a href="http://mbostock.github.io/d3/talk/20111018/collision.html">http://mbostock.github.io/d3/talk/20111018/collision.html</a></code></div>

<aside>
I&rsquo;m using <a href="http://d3js.org">D3</a>, a javascript
library for data visualization.  (D3 stands for &ldquo;data-driven
documents.&rdquo;)<br><br>

This is one of the images that convinced me that I wanted to learn D3.
There are lots of additional examples at the <a
href="https://github.com/mbostock/d3/wiki/Gallery">D3
gallery</a>.<br><br>

There&rsquo;s a ton to learn, and it&rsquo;s rather forbidding (and much of it is
dull).<br><br>

My most common mistake is to forget to include <code>.enter()</code>.
No error is given, but nothing shows up in the plot.
</aside>
<p class="pg">10</p></section>
<script type="text/javascript" src="JS/collision.js"></script>






<section>
<h3><a href="http://jsfiddle.net">JSFiddle</a> is awesome!</h3>

<img height=700 style="position:fixed;bottom:50px;left:100px;" src="Figs/jsfiddle.png">

<aside>
D3 can be forbidding, but just start making stuff. It&rsquo;s fun, and
eventually it&rsquo;s productive.<br><br>

Learn to use your browser&rsquo;s developer tools. When something goes
wrong, typically you just get a blank screen. The developer tools show
the error messages, and you can test things on-the-fly.<br><br>

<a href="http://jsfiddle.net">jsfiddle</a> is a huge help for learning
javascript and D3. There are separate boxes for html, css, javascript,
and the result.
</aside>
<p class="pg">12</p></section>





<section>
<h3><a href="http://coffeescript.org">Coffeescript</a> is awesome!</h3>

<div id="coffee" class="medtop">
<p><code>blob.attr("x", <font class="highlight">(d) -> d.x</font>)</code><br><br>
<code>height = options<font class="highlight">?.</font>height ? 500</code>
</p>
</div>

<aside>
(I think maybe I&rsquo;ve seen the <a
href="http://www.lego.com/en-us/movie">Lego Movie</a> too many
times.)<br><br>

I&rsquo;m using <a href="http://coffescript.org">CoffeeScript</a>
rather than javascript. It gets translated to javascript, and
it&rsquo;s a much nicer language.<br><br>

Particularly nice are the light-weight anonymous
functions. You&rsquo;ll be writing <em>a lot</em> of anonymous
functions in D3.<br><br>

Also, there&rsquo;s a simple syntax for handling <code>null</code>
values, for things like options.<br><br>

You can use <a href="http://jsfiddle.net">jsfiddle</a> with
CoffeeScript (see the next slide).
</aside>
<p class="pg">13</p></section>




<section>
<h3><a href="http://coffeescript.org">Coffeescript</a> is awesome!</h3>

<div id="coffee" class="medtop">
<p><code>blob.attr("x", <font class="highlight">(d) -> d.x</font>)</code><br><br>
<code>height = options<font class="highlight">?.</font>height ? 500</code>
</p>
</div>

<img height=600 style="position:fixed;bottom:10px;left:200px;" src="Figs/jsfiddle_coffee.png">

<aside>
(I think maybe I&rsquo;ve seen the <a
href="http://www.lego.com/en-us/movie">Lego Movie</a> too many
times.)<br><br>

I&rsquo;m using <a href="http://coffescript.org">CoffeeScript</a>
rather than javascript. It gets translated to javascript, and
it&rsquo;s a much nicer language.<br><br>

Particularly nice are the light-weight anonymous
functions. You&rsquo;ll be writing <em>a lot</em> of anonymous
functions in D3.<br><br>

Also, there&rsquo;s a simple syntax for handling <code>null</code>
values, for things like options.<br><br>

You can use <a href="http://jsfiddle.net">jsfiddle</a> with
CoffeeScript (see the next slide).
</aside>
<p class="pg">13</p></section>




<section>
<h3>Interactive genome scan</h3>

<div id="lod_and_effect" class="fig"></div>

<aside>
I&rsquo;ve come to the conclusion that <em>every</em> graph could be
improved with interactivity. You at least want tooltips. Interactive
graphs are incredibly valuable educational tools.<br><br>

The curves at the top are the same LOD curves we saw before. The LOD
score is a measure of association between genotype and phenotype.<br><br>

Click on a chromosome in the top panel to see a detailed view in the
lower-left panel.  Click on a marker in the lower-left panel to see
effect estimates and raw data on the right.  Hover over things on the
right to get some further information.
</aside>

<p class="codelink">
<a href="Coffee/lod_and_effect.coffee" class="gray">code</a> |
<a href="Data/insulinlod.json" class="gray">data</a></p>

<script type="text/javascript" src="JS/lod_and_effect.js"></script>
<p class="pg">14</p></section>





<section nopadding>

<h3 style="margin-top:10px;margin-left:-20px;">Interactive eQTL plot</h3>

<div id="cistrans" class="fig" style="margin-top:-80px;"></div>

<aside>
Here&rsquo;s my most fancy plot. But when I showed this to my
collaborator, his first question was, &ldquo;Where&rsquo;s my favorite
gene?&rdquo; I needed to add a search box. <em>Moral</em>: These tools
are not just for playing around; they need to address the scientific
questions.<br><br>

The top-left image shows inferred eQTL (sites in genome where genotype
is associated with the expression of some gene).  The y-axis
corresponds to the genomic location of probes on a gene expression
array.  The x-axis corresponds to marker positions.  Dots indicate
that the genotypes at a particular marker are associated with the
expression of a particular probe.<br><br>

Hover over a point in the top-right to see all eQTL for a given probe.
Click on a point to see the LOD curves across the genome for that
probe: the association between that probe&rsquo;s expression and genotype at
each position in the genome.<br><br>

Click on a marker in the lower LOD curve plot to see the corresponding
phenotype-vs-genotype plot in the top-right.
</aside>

<p class="codelink">
<a href="Coffee/cistrans.coffee" class="gray">code</a></p>

<script type="text/javascript" src="JS/cistrans.js"></script>
<p class="pg">15</p></section>



<section>
<h3>Write functions</h3>

<ul class="tight">
  <li>One tends to write <font class="highlight">scripts</font></li>
    <ul class="lowlight">
          <li>But they&rsquo;re hard to reuse</li>
    </ul>
  <li>Modularity is the key to reuse</li>
  <li>Need a system for generating plots</li>
  <li>I&rsquo;m working on an <a href="http://www.r-project.org" class="highlight">R</a>
  package, <a href="http://kbroman.github.io/qtlcharts" class="highlight">R/qtlcharts</a></li>
     <ul class="lowlight">
         <li>Functions for different types of panels (scatterplot, heatmap, ...)</li>
         <li>Functions for interactive multi-panel charts</li>
         <li>R functions to generate a chart</li>
     </ul>
</ul>     

<aside>
Rather than write a script with a bunch of code for a specific chart
with specific data, write reusable functions<br><br>

Develop a system for generating interactive charts, so that others
(such as yourself three months from now) can make use of what you&rsquo;ve
built.<br><br>

My <a href="http://kbroman.github.io/qtlcharts">R/qtlcharts</a>
package writes html (including the data, in <a
href="http://www.json.org/">JSON</a> format, and links to the
javascript code) to a file and then opens
it in a browser. For small-to-medium-scale charts, 
this is fast and effective.
</aside>
<p class="pg">16</p></section>






<section>
<h3>Use big selection regions</h3>

<div class="medtop">
<p>And for presentations, bring a mouse!</p>
</div>

<div id="selection_example" style="position:fixed;bottom:0px;left:-50px;">
</div>

<script type="text/javascript" src="JS/selection_example.js"></script>

<aside>
I wanted to point out a few other things I&rsquo;ve learned.<br><br>

First, it can be incredibly frustrating to try to hover over tiny
elements, especially if your hands are shaky.<br><br>

Hover over a point to have it highlighted in pink. The blue dots are
padded, and so easier to select. The small green dots are really hard
to deal with.
</aside>
<p class="pg">17</p></section>






<section>
<h3>A few more things I&rsquo;ve learned</h3>

<ul class="tight">
  <li>For a demo on a laptop at a poster session, ask for a<br>
  <font class="lowlight">&ldquo;cabaret table&rdquo;</font> (like 1m tall).</li>

  <li>Data structures are really important.</li>
     <ul class="lowlight">
       <li>Ill-formed data &rarr; complex code</li>
       <li>Organize by data point rather than by variable</li>
     </ul>

  <li>Testing graphing functions is <font
  class="highlight">hard</font>.</li>
</ul>     

<aside>
In a poster session, a live demo is super, but you don&rsquo;t want people
to have to squat down to use your laptop.<br><br>

I&rsquo;d seldom put much thought into data structures, but in code for
interactive graphs, the layout of the data makes a huge difference.<br><br>

I&rsquo;ve not figured out a good way to make unit tests for code that
produces interactive graphs. I pop up a bunch of examples in a
browser, but then I have to check them manually.
</aside>
<p class="pg">18</p></section>






<section>
<h3>Summary</h3>

<ul class="tight">
  <li><font class="highlight">Interactive</font> graphics
  require effort, but they</li>

      <ul class="lowlight">
        <li>Facilitate exploration</li>
        <li>Are great collaborative tools</li>
        <li>Enable summaries with access to the details</li>
        <li>Are great for teaching</li>
      </ul>

  <li>Visualizations must be <font
  class="highlight">tailored</font> to the data and
  questions</li>

  <li><a href="http://d3js.org" class="highlight">D3</a> is rather low level, but it</li>
      <ul class="lowlight">
        <li>Is totally flexible</li>
        <li>Provides hours of enjoyment</li>
        <li>Can provide other hours of frustration</li>
      </ul>

  <li>Write modular, reusable tools rather than scripts</li>

</ul>

<aside>
It&rsquo;s always good to have a summary slide with your key points.
</aside>
<p class="pg">19</p></section>






<section>
<h3>Acknowledgments<font class="lowlight">: Data</font></h3>

<div style="margin-left:120px;font-size:28px;">

<table cellspacing=25>
  <tr valign="top">
     <td class="lowlight">Alan Attie<br>Mark Keller</td>
     <td width=10></td>
     <td>Biochemistry, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Brian Yandell</td>
     <td></td>
     <td>Statistics and Horticulture, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Christina Kendziorski<br>Aimee Teo Broman</td>
     <td></td>
     <td>Biostatistics & Medical Informatics, UW&ndash;Madison</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Eric Schadt
     <td></td>
     <td>Mount Sinai</td>
     </tr>

  <tr valign="top">
     <td class="lowlight">Danielle Greenawalt<br>Amit Kulkarni</td>
     <td></td>
     <td>Merck & Co., Inc.</td>
     </tr>
</table>
</div>

<aside>
The data were drawn from a study on obesity-dependent diabetes;
many people were involved in that study.
</aside>
<p class="pg">20</p></section>




<section>
<h3>Acknowledgments<font class="lowlight">: Code</font></h3>

<div class="indent">
  <p><a href="http://d3js.org" class="highlight">D3</a> and <a href="https://github.com/mbostock/stack" class="highlight">stack.js</a> by <a href="http://bost.ocks.org/mike" rel="author">Mike Bostock</a>

  <p><a href="http://alignedleft.com/tutorials/d3"
  class="highlight">D3 tutorials</a> by <a href="http://alignedleft.com">Scott Murray</a>

  <p><a href="http://coffeescript.org" class="highlight">CoffeeScript</a> by <a
  href="https://github.com/jashkenas">Jeremy Ashkenas</a>

  <p><a href="http://eloquentjavascript.net"
  class="highlight"><i>Eloquent JavaScript</i></a> by <a href="http://marijnhaverbeke.nl">Marijn Haverbeke</a>

  <p><a href="http://pragprog.com/book/tbcoffee/coffeescript"
  class="highlight"><i>CoffeeScript</i> book</a> by <a href="http://trevorburnham.com">Trevor Burnham</a>

  <p><a href="https://github.com/Caged/d3-tip" class="highlight">d3-tip</a> by <a href="http://labratrevenge.com">Justin Palmer</a>
</div>

<aside>
The interactive examples
were produced with the <a href="http://d3js.org">D3</a>
library (written in JavaScript).<br><br>

I wrote the code in CoffeeScript,
which is translated to JavaScript but is a better-designed
language.<br><br>

<a
href="http://eloquentjavascript.net"><i>Eloquent JavaScript</i></a> is
a great free book.  The <a
href="http://pragprog.com/book/tbcoffee/coffeescript"><i>CoffeeScript</i></a>
book is not free but is also great; you need to learn at least a bit of
JavaScript first.<br><br>

<a href="http://alignedleft.com/tutorials/d3">Scott
Murray&rsquo;s D3 tutorials</a> are superb and are a great place to start
for learning D3; he&rsquo;s since written <a
href="http://www.amazon.com/exec/obidos/ASIN/1449339735/7210-20">a
book</a>.<br><br>

These slides are in html, using <a
href="https://github.com/mbostock/stack">stack.js</a>.  I used <a
href="http://github.com/kbroman/d3-tip">a modified version</a> of <a
href="http://github.com/Caged/d3-tip">d3-tip</a> for tool tips.
</aside>
<p class="pg">21</p></section>






<section>

<div class="hugetop">
<p class="large"><code>
<a href="http://www.biostat.wisc.edu/~kbroman" class="gray">biostat.wisc.edu/~kbroman</a><br>
</code></p>

<p class="large"><code>
<a href="https://github.com/kbroman" class="gray">github.com/kbroman</a><br>
</code></p>

<p class="large"><code>
<a href="https://twitter.com/kwbroman" class="gray">@kwbroman</a>
</code></p>

<p class="large"><code>
<a href="https://github.com/kbroman/qtlcharts" class="gray">bit.ly/qtlcharts</a>
</code></p>
</div>

<aside>
This is where to find me.<br><br>

In particular, you can find all of my papers, talks, and software.
</aside>
<p class="pg">22</p></section>



<script type="text/javascript" src="JS/stack.v0.js"></script>
</body>
</html>

